File: sdks/typescript/src/claims.ts
/**
 * Spec §4 — Claims API.
 *
 * {@link Claims} wraps a decoded JWT payload and exposes typed accessors
 * for standard OIDC and Hearth-specific claims.  All reads are local —
 * no network call is made.  Signature verification is the caller's
 * responsibility (e.g. via the JWKS endpoint before constructing Claims).
 */

import { decodeJwt } from "jose";
import {
  TokenExpiredError,
  TokenInvalidError,
  TokenNotYetValidError,
} from "./errors.js";

/** Raw JWT payload shape used internally. */
interface RawPayload {
  sub?: string;
  iss?: string;
  aud?: string | string[];
  exp?: number;
  nbf?: number;
  iat?: number;
  jti?: string;
  scope?: string;
  scopes?: string[];
  roles?: string[];
  permissions?: string[];
  [key: string]: unknown;
}

/**
 * Typed accessor for a decoded JWT's claims.
 *
 * Construct via {@link Claims.decode} (decodes without verifying signature)
 * or pass a pre-decoded payload to the constructor.
 */
export class Claims {
  private readonly payload: RawPayload;

  constructor(payload: RawPayload) {
    this.payload = payload;
  }

  /**
   * Decode a JWT string into a {@link Claims} object.
   * The signature is NOT verified — the caller must verify it separately.
   *
   * @throws {TokenInvalidError} if the string is not a valid JWT.
   */
  static decode(token: string): Claims {
    try {
      const payload = decodeJwt(token) as RawPayload;
      return new Claims(payload);
    } catch (err) {
      throw new TokenInvalidError(
        `Failed to decode JWT: ${err instanceof Error ? err.message : String(err)}`,
      );
    }
  }

  /**
   * Assert the token is temporally valid (not expired, past nbf).
   *
   * @throws {TokenExpiredError} if exp is in the past.
   * @throws {TokenNotYetValidError} if nbf is in the future.
   */
  assertValid(clockSkewSeconds = 0): void {
    const now = Math.floor(Date.now() / 1000);
    const exp = this.payload.exp;
    if (exp !== undefined && now > exp + clockSkewSeconds) {
      throw new TokenExpiredError(new Date(exp * 1000));
    }
    const nbf = this.payload.nbf;
    if (nbf !== undefined && now < nbf - clockSkewSeconds) {
      throw new TokenNotYetValidError(new Date(nbf * 1000));
    }
  }

  /** The `sub` (subject) claim — identifies the principal that is the subject of the JWT. */
  subject(): string {
    return this.payload.sub ?? "";
  }

  /** The `iss` (issuer) claim — identifies the principal that issued the JWT. */
  issuer(): string {
    return this.payload.iss ?? "";
  }

  /** The `aud` (audiences) claim — normalized to an array. */
  audiences(): string[] {
    const aud = this.payload.aud;
    if (!aud) return [];
    return Array.isArray(aud) ? aud : [aud];
  }

  /** The `exp` (expiry) claim as a Date, or null if absent. */
  expiry(): Date | null {
    return this.payload.exp !== undefined
      ? new Date(this.payload.exp * 1000)
      : null;
  }

  /** The `iat` (issuedAt) claim as a Date, or null if absent. */
  issuedAt(): Date | null {
    return this.payload.iat !== undefined
      ? new Date(this.payload.iat * 1000)
      : null;
  }

  /** The `jti` (JWT ID) claim, or null if absent. */
  jwtID(): string | null {
    return this.payload.jti ?? null;
  }

  /** The `scope` claim split into individual scopes (or `scopes` array if present). */
  scopes(): string[] {
    if (this.payload.scopes) return this.payload.scopes;
    const scope = this.payload.scope;
    if (!scope) return [];
    return scope.split(/\s+/).filter(Boolean);
  }

  /** Returns true iff the token contains the given scope. */
  hasScope(scope: string): boolean {
    return this.scopes().includes(scope);
  }

  /** Returns true iff the token's `roles` claim contains the given role. */
  hasRole(role: string): boolean {
    return (this.payload.roles ?? []).includes(role);
  }

  /** Returns true iff the token's `permissions` claim contains the given permission. */
  hasPermission(permission: string): boolean {
    return (this.payload.permissions ?? []).includes(permission);
  }

  /** Access an arbitrary claim by key. */
  get(key: string): unknown {
    return this.payload[key];
  }
}

File: sdks/rust/src/claims.rs
//! Spec §4 — Claims API.
//!
//! [`Claims`] wraps a decoded JWT payload and exposes typed accessors
//! for standard OIDC and Hearth-specific claims.  All reads are local —
//! no network call is made.  Signature verification is the caller's
//! responsibility.

use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine as _};
use serde_json::Value;

use crate::error::HearthError;

/// Typed accessor for a decoded JWT's claims (spec §4).
///
/// Construct via [`Claims::decode`] or pass a pre-decoded [`serde_json::Value`]
/// to [`Claims::from_value`].
pub struct Claims {
    payload: Value,
}

impl Claims {
    /// Decode a JWT string without verifying its signature.
    ///
    /// # Errors
    /// Returns [`HearthError::TokenInvalidError`] if the string is not a
    /// structurally valid JWT or the payload segment cannot be decoded.
    pub fn decode(token: &str) -> Result<Self, HearthError> {
        let parts: Vec<&str> = token.split('.').collect();
        if parts.len() != 3 {
            return Err(HearthError::TokenInvalidError {
                reason: "expected three dot-separated segments".into(),
            });
        }
        let payload_bytes = URL_SAFE_NO_PAD
            .decode(parts[1])
            .map_err(|e| HearthError::TokenInvalidError {
                reason: format!("base64 decode: {e}"),
            })?;
        let payload: Value =
            serde_json::from_slice(&payload_bytes).map_err(|e| HearthError::TokenInvalidError {
                reason: format!("JSON parse: {e}"),
            })?;
        Ok(Self { payload })
    }

    /// Construct from a pre-decoded JSON value.
    pub fn from_value(payload: Value) -> Self {
        Self { payload }
    }

    /// Assert the token is temporally valid.
    ///
    /// # Errors
    /// - [`HearthError::TokenExpiredError`] when `exp` is in the past.
    /// - [`HearthError::TokenNotYetValidError`] when `nbf` is in the future.
    pub fn assert_valid(&self, clock_skew_seconds: i64) -> Result<(), HearthError> {
        use std::time::{SystemTime, UNIX_EPOCH};
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_secs() as i64)
            .unwrap_or(0);

        if let Some(exp) = self.payload.get("exp").and_then(|v| v.as_i64()) {
            if now > exp + clock_skew_seconds {
                return Err(HearthError::TokenExpiredError { expired_at: exp });
            }
        }
        if let Some(nbf) = self.payload.get("nbf").and_then(|v| v.as_i64()) {
            if now < nbf - clock_skew_seconds {
                return Err(HearthError::TokenNotYetValidError { not_before: nbf });
            }
        }
        Ok(())
    }

    // ── Spec §4 accessor methods ─────────────────────────────────────────
    // Method names follow the spec identifiers exactly; allow non_snake_case
    // for camelCase names required by the spec surface.
    #[allow(non_snake_case)]

    /// Return the `sub` (subject) claim.
    pub fn subject(&self) -> &str {
        self.payload
            .get("sub")
            .and_then(|v| v.as_str())
            .unwrap_or("")
    }

    /// Return the `iss` (issuer) claim.
    pub fn issuer(&self) -> &str {
        self.payload
            .get("iss")
            .and_then(|v| v.as_str())
            .unwrap_or("")
    }

    /// Return the `aud` (audiences) claim normalised to a `Vec`.
    pub fn audiences(&self) -> Vec<String> {
        match self.payload.get("aud") {
            None => vec![],
            Some(Value::String(s)) => vec![s.clone()],
            Some(Value::Array(arr)) => arr
                .iter()
                .filter_map(|v| v.as_str().map(str::to_string))
                .collect(),
            _ => vec![],
        }
    }

    /// Return the `exp` claim as a Unix timestamp, or `None` if absent.
    pub fn expiry(&self) -> Option<i64> {
        self.payload.get("exp").and_then(|v| v.as_i64())
    }

    /// Return the `iat` claim as a Unix timestamp, or `None` if absent.
    #[allow(non_snake_case)]
    pub fn issuedAt(&self) -> Option<i64> {
        self.payload.get("iat").and_then(|v| v.as_i64())
    }

    /// Return the `jti` (JWT ID) claim, or `None` if absent.
    #[allow(non_snake_case)]
    pub fn jwtID(&self) -> Option<&str> {
        self.payload.get("jti").and_then(|v| v.as_str())
    }

    /// Return the individual scopes from the `scope` claim.
    pub fn scopes(&self) -> Vec<String> {
        self.payload
            .get("scope")
            .and_then(|v| v.as_str())
            .map(|s| s.split_whitespace().map(str::to_string).collect())
            .unwrap_or_default()
    }

    /// Return `true` iff the token contains the given scope.
    #[allow(non_snake_case)]
    pub fn hasScope(&self, scope: &str) -> bool {
        self.scopes().iter().any(|s| s == scope)
    }

    /// Return `true` iff the token's `roles` claim contains the given role.
    #[allow(non_snake_case)]
    pub fn hasRole(&self, role: &str) -> bool {
        self.payload
            .get("roles")
            .and_then(|v| v.as_array())
            .map(|arr| arr.iter().any(|v| v.as_str() == Some(role)))
            .unwrap_or(false)
    }

    /// Return `true` iff the token's `permissions` claim contains the given permission.
    #[allow(non_snake_case)]
    pub fn hasPermission(&self, permission: &str) -> bool {
        self.payload
            .get("permissions")
            .and_then(|v| v.as_array())
            .map(|arr| arr.iter().any(|v| v.as_str() == Some(permission)))
            .unwrap_or(false)
    }

    /// Return an arbitrary claim by key.
    pub fn get(&self, key: &str) -> Option<&Value> {
        self.payload.get(key)
    }
}
